### ISSUE TITLE
[2020/10/21] Sanity-Checking Pruning Methods: Random Tickets can Win the Jackpot

### ISSUE THREAD

**Issue by zaemyung** [zaemyung/Seeking-SOTA#5]
- Sanity-Checking Pruning Methods: Random Tickets can Win the Jackpot
- Neurips 2020
- https://arxiv.org/pdf/2009.11094.pdf

**Comment by kwangheeya** [comment]
i wanna Neuritis

**Comment by zaemyung** [comment]
@lekwanghee @passing2961 @hagg30 
아침에 또 집 보러 가는 미팅이 생겨서, 두 시간 반만 땡길 수 있을까요: **한국시간 오후 2:00**

**Comment by passing2961** [comment]
> @lekwanghee @passing2961 @hagg30
> 아침에 또 집 보러 가는 미팅이 생겨서, 두 시간 반만 땡길 수 있을까요: **한국시간 오후 2:00**

네 저는 가능합니다~

**Comment by zaemyung** [comment]
Comparing Rewinding and Fine-tuning in Neural Network Pruning
https://openreview.net/forum?id=S1gSj0NKvB
해당 논문에서 베이스라인으로 비교하고 있는 위 논문도 자매품으로 같이 읽으면 좋을 듯 합니다. Lottery ticket hypothesis 저자의 follow-up 논문입니다 (ICLR 2020).

**Comment by zaemyung** [comment]
## Pruning
- Removing weights, filters to compress the original networks for inference speed, storage consumption, energy consumption, etc.
- Unstructured pruning: prune individual weights
- Structured pruning: prune weights in groups, convolutional filters, channels, etc.

### Methods
- Regularization
- Gradual pruning
- **Retraining**
  - TRAIN the network to completion
  - PRUNE structures of the network according to some heuristics
  - RE-TRAIN the subnetwork
  - (REPEAT until desired sparsity) or do ONE-SHOT

## "Comparing Rewinding and Fine-tuning in Neural Network Pruning"
### The lottery ticket hypothesis
> Early in training, sparse subnetworks emerge that can train in isolation to the same accuracy as the original network.
- We can find these "tickets" (subnetworks) via **weight rewinding** and **learning rate rewinding**.
- Weight rewinding
  - After the TRAIN and PRUNE steps, rewind the unpruned weights to their values from _t_ epochs earlier in training.
    It also rewinds the learning rate schedule as well.
- Learning rate rewinding (new in _"Comparing Rewinding and Fine-tuning in Neural Network Pruning"_)
  - Only rewind the learning rate schedule to _t_ epochs earlier. Use the final weights from the training.
- c.f. rewinding safe zone
- Result: learning rate rewinding >= weight rewinding > fine-tuning
- In the paper, pruning is done globally, i.e. 20% of the weights are pruned based on their magnitude.

## "Sanity-Checking Pruning Methods: Random Tickets can Win the Jackpot"
- Empirically shows that the following two common beliefs on pruning are possibly incorrect:
  1) "Pruning methods exploit the information from training data in the pruning step to finding good subnetworks."
  2) "The architecture of the pruned network is crucial for obtaining the final pruned network with good performance."

### Preliminaries

![image](https://user-images.githubusercontent.com/3746478/96660743-aca60080-134a-11eb-99d4-e59ebb505a0b.png)

### Sanity-checking the first belief
- Random labels
- Random pixels
- Half dataset

### Sanity-checking the second belief
- Layer-wise rearrangement: randomly rearrange the mask obtained from pruning per layer
- Layer-wise weights shuffling: keep the mask, but only change the unmasked weights

![image](https://user-images.githubusercontent.com/3746478/96661621-cba59200-134c-11eb-8eb7-b623ad46f9f8.png)

![image](https://user-images.githubusercontent.com/3746478/96660803-d95a1800-134a-11eb-9418-6dfbca4f1105.png)

### Results
#### Initial Tickets Fail to Pass Sanity Checks
![image](https://user-images.githubusercontent.com/3746478/96660900-17573c00-134b-11eb-9fa1-2184bc2fbf38.png)

- The results of sanity check on initial tickets suggest that:
  - Data may not be important for finding good initial tickets.
  - The connections of neurons in individual layers in initial tickets can be totally rearranged without any performance drop, indicating that only the number of remained connections or **keep-ratio** in each layer matters.

#### Random Tickets Win the Jackpot
- From observation on SNIP, GraSP, the authors found that the keep-ration should decrease with the depth of the network.
- "Smart ratio": set keep-ration proportional to `(L-l+1)^2 + (L-l+1)`.

![image](https://user-images.githubusercontent.com/3746478/96661278-fba06580-134b-11eb-9cd9-0a8bd5cc67c6.png)

#### Partially-trained Tickets (learning rate rewinding) Pass Sanity Checks
- Hybrid tickets: `smart ratio + learning rate rewinding`
  > We first pre-train the network as in learning rate rewinding, and then, while pruning the pre-trained network, we use magnitude-based pruning layerwise and the keep-ratios are determined by the smart-ratios. In other words, for a given layer l, we only keep those weights whose magnitude is in the largest pl portion in layer l, where pl is the smart-ratio of layer l. And finally, the pruned network is retrained with a full learning-rate schedule as learning rate rewinding does

![image](https://user-images.githubusercontent.com/3746478/96661410-4326f180-134c-11eb-81f3-3519fad721c4.png)
